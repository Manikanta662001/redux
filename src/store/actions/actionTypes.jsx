const INCREMENT = "INCREMENT";
const DECREMENT = "DECREMENT";
const RESET = "RESET";
const INCREMENT_BY_VALUE = "INCREMENT_BY_VALUE";
const FETCH_USERS_REQUEST = "FETCH_USERS_REQUEST";
const FETCH_USERS_SUCCESS = "FETCH_USERS_SUCCESS";
const FETCH_USERS_FAILURE = "FETCH_USERS_FAILURE";

export {
  INCREMENT,
  DECREMENT,
  RESET,
  INCREMENT_BY_VALUE,
  FETCH_USERS_REQUEST,
  FETCH_USERS_SUCCESS,
  FETCH_USERS_FAILURE,
};
