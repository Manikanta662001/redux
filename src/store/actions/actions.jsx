import {
  DECREMENT,
  FETCH_USERS_FAILURE,
  FETCH_USERS_REQUEST,
  FETCH_USERS_SUCCESS,
  INCREMENT,
  INCREMENT_BY_VALUE,
  RESET,
} from "./actionTypes";

export const increment = () => {
  return { type: INCREMENT };
};
export const decrement = () => {
  return { type: DECREMENT };
};
export const reset = () => {
  return { type: RESET };
};
export const increment_by_async = () => {
  return { type: INCREMENT };
};
export const increment_by_value = (val) => {
  return { type: INCREMENT_BY_VALUE, payload: val };
};

export const fetchUsersRequest = () => {
  return { type: FETCH_USERS_REQUEST };
};
export const fetchUsersSuccess = (data) => {
  return { type: FETCH_USERS_SUCCESS, payload: data };
};
export const fetchUsersFailure = (errmsg) => {
  return { type: FETCH_USERS_FAILURE, payload: errmsg };
};

export const fetchUsers = () => {
  return (dispatch) => {
    dispatch(fetchUsersRequest());
    fetch("https://jsonplaceholder.typicode.com/users")
      .then((res) => res.json())
      .then((result) => dispatch(fetchUsersSuccess(result)))
      .catch((err) => dispatch(fetchUsersFailure(err.message)));
  };
};
