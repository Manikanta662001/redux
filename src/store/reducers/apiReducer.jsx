import {
  FETCH_USERS_FAILURE,
  FETCH_USERS_REQUEST,
  FETCH_USERS_SUCCESS,
} from "../actions/actionTypes";

const initialState = {
  users: [],
  isLoading: false,
  errormsg: "",
};
const apiReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_USERS_REQUEST:
      return { ...state, isLoading: true };
    case FETCH_USERS_SUCCESS:
      return { ...state, isLoading: false, users: action.payload ,errormsg:""};
    case FETCH_USERS_FAILURE:
      return { ...state, isLoading: false, errormsg: action.payload ,users:[]};
    default:
      return state;
  }
};

export default apiReducer;
