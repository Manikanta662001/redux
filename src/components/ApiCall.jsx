import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchUsers } from "../store/actions/actions";

function ApiCall({url}) {
  const users = useSelector((state) => state.apiReducer.users);
  const error = useSelector((state) => state.apiReducer.errormsg);
  const dispatch = useDispatch();
  console.log(users, error);
  return (
    <div>
      <button onClick={() => dispatch(fetchUsers(url))}>Api call</button>
      {error === ""
        ? users?.map((user) => <li key={user.id}>{user.name}</li>)
        : error}
    </div>
  );
}

export default ApiCall;
