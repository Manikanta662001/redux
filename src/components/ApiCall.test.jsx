import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { screen, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import apiReducer from "../store/reducers/apiReducer";
import ApiCall from "./ApiCall";
import { Provider } from "react-redux";
import { fetchUsers } from "../store/actions/actions";
const rootReducer = combineReducers({
  apiReducer,
});
const renderWithProvider = (component, { initialState } = {}) => {
  const store = configureStore({
    reducer: rootReducer,
    preloadedState: initialState,
  });
  return { ...render(<Provider store={store}>{component}</Provider>), store };
};

jest.mock("../store/actions/actions", () => ({
  fetchUsers: jest.fn(),
}));

describe("ApiCall Component", () => {
  const initialState = {
    apiReducer: { users: [], isLoading: false, errormsg: "" },
  };
  const user = userEvent.setup();
  test("renders correctly and Match the Snapshot", () => {
    const { asFragment } = renderWithProvider(
      <ApiCall url={"https://jsonplaceholder.typicode.com/users"} />,
      { initialState }
    );
    expect(
      asFragment(<ApiCall url={"https://jsonplaceholder.typicode.com/users"} />)
    ).toMatchSnapshot();
  });
  test("fetching users without error", async () => {
    fetchUsers.mockImplementation(() => (dispatch) => {
      dispatch({
        type: "FETCH_USERS_SUCCESS",
        payload: [{ id: 1, name: "John Doe" }],
      });
    });
    renderWithProvider(
      <ApiCall url={"https://jsonplaceholder.typicode.com/users"} />,
      { initialState }
    );
    const apiBtn = screen.getByRole("button", { name: "Api call" });
    await user.click(apiBtn);
    const litags = await screen.findAllByRole("listitem");
    expect(litags).toHaveLength(1);
    expect(litags[0]).toHaveTextContent("John Doe");
  });
  test("error while fetching", async () => {
    fetchUsers.mockImplementation(() => (dispatch) => {
      dispatch({
        type: "FETCH_USERS_FAILURE",
        payload: "Error while Fetching",
      });
    });
    renderWithProvider(
      <ApiCall url={"https://jsonplaceholder.typicode.com/users"} />,
      { initialState }
    );
    const apiBtn = screen.getByRole("button", { name: "Api call" });
    await user.click(apiBtn);
    expect(await screen.findByText(/Error while Fetching/)).toBeInTheDocument();
  });
});
