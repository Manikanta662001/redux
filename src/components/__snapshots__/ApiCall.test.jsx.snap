// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiCall Component renders correctly and Match the Snapshot 1`] = `
<DocumentFragment>
  <div>
    <button>
      Api call
    </button>
  </div>
</DocumentFragment>
`;
